### PR TITLE
fix: appWithTranslation useMemo bug (null locale)

### LINKED ISSUE #1628: Next/Link causing app to unmount
## Describe the bug
For some reason if you use a next/link component and click the same link twice, the entire app will unmount and mount itself, causing some parts of my application to behave weirdly. Upon further inspection, clicking a link your already on will set the locale key on App to null (which is probably what's causing the issue). It seems that this only happens in production (im using vercel). 

**next-i18next.config.js**
`module.exports = {
  i18n: {
    defaultLocale: "en-US",
    locales: ["en-US", "sv-SE"],
  },
};
`

## Occurs in next-i18next version
10.1.0

## Steps to reproduce
I have created a minimal setup [repo](https://github.com/JonahLund/next-i18next_link_bug) that's deployed to [vercel](https://next-i18next-link-bug.vercel.app/)

## Expected behaviour
The app should not unmount if you click on a next/link component twice.

## Screenshots
Everything fine
image
Until you click on the same link again
image

### PR DESCRIPTION
**Pull request by JonahLund**
This PR will fix the Next/Link app unmount.

I placed the locale in a state instead of a variable, which fixes locale getting set to null.

The state will only change if the `initialLocale` has changed and is not equal to the current state and is not null.

Fixes #1628

### CODE REVIEW

File: src/appWithTranslation.tsx [hunk 1]
```diff
@@ -23,7 +23,7 @@ export const appWithTranslation = <Props extends AppProps = AppProps>(
 ) => {
   const AppWithTranslation = (props: Props) => {
     const { _nextI18Next } = props.pageProps as SSRConfig
-    let locale = null
+    const [locale, setLocale] = useState<string>(_nextI18Next?.initialLocale)
```
Comment (isaachinman) [inline src/appWithTranslation.tsx:26]:
```suggestion
    const [locale, setLocale] = useState<string>(_nextI18Next?.initialLocale ?? null)
```

**Review by isaachinman — COMMENTED** [review]
Thanks! Do you mind writing some test coverage?

**Comment by JonahLund** [comment]
I think i came up with a better and simpler solution that doesnt require any state hooks

**Comment by JonahLund** [comment]
Good thing those tests were there

**Comment by JonahLund** [comment]
> Thanks! Do you mind writing some test coverage?

I assume that the test coverage isn't needed now that it passed?

**Comment by JonahLund** [comment]
It seems that there were already a PR addressing this issue #1548

**Comment by wulimao49** [comment]
there is a workaround
`memo(appWithTranslation(MyApp))`

**Comment by isaachinman** [comment]
@JonahLund I didn't get a response from @wulimao49 on #1548, so wasn't sure how to proceed. In any case, I think the solution here is simpler, so let's proceed with this PR.

Can we just add basic tests that pass `_nextI18Next?.initialLocale` as some expected value, and assert that `locale` gets set to it?

Specifically it would be great if we could actually write a test that reproduces the buggy `null` behaviour we are fixing, to ensure we do not have any regressions in the future.

**Comment by JonahLund** [comment]
> @JonahLund I didn't get a response from @wulimao49 on #1548, so wasn't sure how to proceed. In any case, I think the solution here is simpler, so let's proceed with this PR.
> 
> Can we just add basic tests that pass `_nextI18Next?.initialLocale` as some expected value, and assert that `locale` gets set to it?
> 
> Specifically it would be great if we could actually write a test that reproduces the buggy `null` behaviour we are fixing, to ensure we do not have any regressions in the future.

Any idea of how to do that? I tried writing a test that should fail in the non fixed version but it doesn't seem to do that.

**Comment by isaachinman** [comment]
All it should take is re-rendering `appWithTranslation` with the same `locale` a few times, right?

**Comment by JonahLund** [comment]
> All it should take is re-rendering `appWithTranslation` with the same `locale` a few times, right?

Yeah exactly

**Comment by isaachinman** [comment]
Only thing that comes to mind is that this `useMemo` bug only seems to occur in production mode, so perhaps the default jest env won't work.

**Comment by JonahLund** [comment]
> Only thing that comes to mind is that this `useMemo` bug only seems to occur in production mode, so perhaps the default jest env won't work.

Yeah maybe, i tried this but the key stays 'de', so i dont know if im using the `.mock.calls` wrong or if the `rerender` function doesn't work the way i think it works, im definitely not any professional at jest testing 😂
image

**Comment by isaachinman** [comment]
According to [the docs](https://testing-library.com/docs/example-update-props/), the `rerender` function should work. Do you mind having another go, or sharing what WIP code you have?

**Comment by JonahLund** [comment]
> According to [the docs](https://testing-library.com/docs/example-update-props/), the `rerender` function should work. Do you mind having another go, or sharing what WIP code you have?

Yeah so basically we just need to access the key prop to verify that it's not null. And im not entirely sure how to do that via the react testing library

**Comment by isaachinman** [comment]
You can add a `data-locale` (or similar) attribute if it helps for testing purposes 👍

**Comment by JonahLund** [comment]
Could also just throw an error if locale is null, and then just make sure it doesnt throw an error in the test.

**Comment by JonahLund** [comment]
Well if _nextI18Next isnt null and locale is*

**Comment by isaachinman** [comment]
If we wanted to throw errors, we would need to bump the major version. Let's just stick with a bug fix for now.

**Comment by JonahLund** [comment]
> Hey @JonahLund – looks like we've got some failing tests.
> 
> This issue is affecting a fair few users, so it'd be great to get it merged as soon as we can.
> 
> Please let me know if you've still got time for this, otherwise I don't mind taking over. Thanks!

Hello, I wasn't able to successfully reproduce the bug in the testing coverage,  so that's why I haven't made any new commits.

Yeah the only thing left is rerendering the component and making sure the `key` prop isn't null. Which is what i was unable to do. So if you are able to reproduce it then that would be great :)

**Comment by JonahLund** [comment]
The bug fix works as it should, just not the latest testing coverage i committed.

**Comment by JonahLund** [comment]
Ok I will give it another try with the tests.

**Comment by JonahLund** [comment]
I tried with a different react testing library, was easier to get all properties from the component, but still can't seem to reproduce the `null` behaviour. This is what i came up with

When running this test on the bugged version, it passes, which it shouldn't.

`it('assures locale key is set to the right value', () => {
    const component = renderer.create(<DummyApp
      {...(createProps('de'))}
    />)

    component.update(<DummyApp
      {...(createProps('en'))}
    />)

    component.update(<DummyApp
      {...(createProps('en'))}
    />)

    const getKeyProp = ((root: any) => root._fiber.child.child.key)

    expect(getKeyProp(component.root)).toBe('en')
  })`

Maybe this is something that has to be e2e tested in cypress, im not sure.

**Comment by isaachinman** [comment]
@JonahLund Thanks again for spending so much time on this. Let me step in and help, as it's Sunday and I have a few minutes. It's absolutely possible to introspect `children.key`, and I believe that's the right approach here. The only thing you were missing is that the `createProps` function returns a new object every time, so memoisation was actually not taking place.

It's a little cumbersome, but we can "trick" memoisation into taking effect by mutating the same object in place:

```tsx
it('assures locale key is set to the right value', () => {
  const props = createProps('de')

  const { rerender } = render(<DummyApp
    {...props}
  />)

  props.router.locale = 'en'
  props.pageProps._nextI18Next.initialLocale = 'en'

  rerender(<DummyApp
    {...props}
  />)
  rerender(<DummyApp
    {...props}
  />)
  rerender(<DummyApp
    {...props}
  />)

  props.router.locale = 'de'
  props.pageProps._nextI18Next.initialLocale = 'de'

  rerender(<DummyApp
    {...(createProps('de'))}
  />)

  const [[first], [second], [third], [fourth], [fifth]] =
    (I18nextProvider as jest.Mock).mock.calls

  expect(first.children.key).toEqual('de')
  expect(second.children.key).toEqual('en')
  expect(third.children.key).toEqual('en')
  expect(fourth.children.key).toEqual('en')
  expect(fifth.children.key).toEqual('de')
})
```

Crucially, note that if you revert the actual source code changes, this test will fail.

If you can add this test coverage, I will merge immediately. Thanks!

**Comment by JonahLund** [comment]
Awesome! :)

**Review by isaachinman — APPROVED** [review]

**Comment by isaachinman** [comment]
Thank you @JonahLund!

**Pull request merged**